vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,876 @@
+
+  size_t global_lease;
+  size_t max_live_configs;
+  std::string cm_type;
+  std::string priority_type;
+  int priority_challengers;
+
+  option_group_definition new_options("Debug: automl reduction");
+  new_options
+      .add(make_option("automl", max_live_configs)
+               .necessary()
+               .keep()
+               .default_value(3)
+               .help("set number of live configs"))
+      .add(make_option("global_lease", global_lease)
+               .keep()
+               .default_value(10)
+               .help("set initial lease for automl interactions"))
+      .add(make_option("cm_type", cm_type).keep().default_value("interaction").help("set type of config manager"))
+      .add(make_option("priority_type", priority_type)
+               .keep()
+               .default_value("none")
+               .help("set function to determine next config {none, least_exclusion}"))
+      .add(make_option("priority_challengers", priority_challengers)
+               .keep()
+               .default_value(-1)
+               .help("set number of priority challengers to use"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) return nullptr;
+
+  priority_func* calc_priority;
+
+  if (priority_type == "none") { calc_priority = &calc_priority_empty; }
+  else if (priority_type == "least_exclusion")
+  {
+    calc_priority = &calc_priority_least_exclusion;
+  }
+  else
+  {
+    THROW("Error: Invalid priority function provided");
+  }
+
+  if (priority_challengers < 0) { priority_challengers = (static_cast<int>(max_live_configs) - 1) / 2; }
+
+  auto cm = VW::make_unique<interaction_config_manager>(
+      global_lease, max_live_configs, all.random_seed, static_cast<size_t>(priority_challengers), calc_priority);
+  auto data = VW::make_unique<automl<interaction_config_manager>>(std::move(cm));
+  // all is not needed but good to have for testing purposes
+  data->all = &all;
+  assert(max_live_configs <= MAX_CONFIGS);
+
+  // override and clear all the global interactions
+  // see parser.cc line 740
+  all.interactions.clear();
+  assert(all.interactions.empty() == true);
+
+  // make sure we setup the rest of the stack with cleared interactions
+  // to make sure there are not subtle bugs
+  auto* base_learner = stack_builder.setup_base_learner();
+
+  assert(all.interactions.empty() == true);
+
+  assert(all.weights.sparse == false);
+  if (all.weights.sparse) THROW("--automl does not work with sparse weights");
+
+  details::fail_if_enabled(all,
+      {"ccb_explore_adf", "audit_regressor", "baseline", "cb_explore_adf_rnd", "cb_to_cb_adf", "cbify", "replay_c",
+          "replay_b", "replay_m", "memory_tree", "new_mf", "nn", "stage_poly"});
+
+  // only this has been tested
+  if (base_learner->is_multiline)
+  {
+    // fetch cb_explore_adf to call directly into the print routine twice
+    data->adf_learner = as_multiline(base_learner->get_learner_by_name_prefix("cb_explore_adf_"));
```

Comment:
does base->print_example not work?  Seems like it would be good to avoid making assumption about the stack if possible.

Reply:
or if the functionality of cb_explore_adf print() is needed then perhaps refer to that function directly?

Reply:
Adding dependency on a very specific reduction to print out progress seems like excessive coupling to me.  It would be less troubling if this was not getting merged to master.  My worry is that this will never get addressed.  Having said that, adding a tech-debt section to the project stories and putting this a story under it seems like a way to move forward.  Let me know if that makes sense and also questions about where/how to add it.